Si ai

Sam-Irv opened the pull request:
AI implementation

Sam-Irv committed:
updated ai.py to compute what the best move is

Sam-Irv committed:
updated ai to use rank_move to determine ai move

Sam-Irv committed:
Corrected the single player with AI test

Adjusted the API call to account for variable numbers of players.

jacobstringer commented on application/controllers/ai.py:
```diff
@@ -1,25 +1,228 @@
+        if card.giveMoney and player.money < 2*card.age:
+            #bonus when low on money
+            score += 2*card.age - player.money
+        cards_holder = [get_cards(player=neighbours[0], history=True), get_cards(player=player, history=True), get_cards(player=neighbours[1], history=True)]
```
Cards_holder never gets used?

jacobstringer commented on application/controllers/ai.py:
```diff
@@ -1,25 +1,228 @@
+    score -= card.costMoney
+    
+    #penalty for losing money
+    trade_dict = check_move_and_trade(card, player)
```
check_move_and_trade is a fairly high time complexity function. I would recommend only calling it once, either here, or in ai_move and then piping the returned object through to rank_move.

jacobstringer commented on application/controllers/ai.py:
```diff
@@ -1,25 +1,228 @@
+        score += 2*card.age - player.money
+    
+    
+    if card.name == 'Marketplace':
```
I would split this if statement in two, one for yellow and purple cards, then before entering the if statement, check if the card is that colour. Currently, every card goes through 18 predicate checks when they may not even be a yellow or purple card.

jacobstringer commented on application/controllers/card_logic.py:
```diff
@@ -55,19 +55,31 @@ def process_card(card, player, is_discarded, for_wonder, from_discard_pile=None,
-        if play_card(card, player, False, False):
-            swap_hands(card, player, game)
-            return
+        if check_valid_move(card, player):
```
check_move_with_trade(card, player) will check everything check_valid_move does but includes trade.

jacobstringer commented on application/controllers/card_logic.py:
```diff
@@ -55,19 +55,31 @@ def process_card(card, player, is_discarded, for_wonder, from_discard_pile=None,
+
+    if best_card is False:
+        #discard at random
+        play_card(cards[0], player, True, False)
```
play_card_with_trade(card, player, is_discarded, for_wonder, trade) replaces the deprecated play_card function.

jacobstringer reviewed (changes requested):
Looks great, works well. Added some comments about performance improvements and moving the AI to the new trade integrated functions.

Sam-Irv committed:
rank_move paramenters adjusted

now requires trade_dict, used to reduce the number of (expensive) trade algorithm calls.

Sam-Irv committed:
updated ai_move

now allows ai to play cards via trade, as well as passing the rank_move function the trade_dict to reduce calls to the (expensive) trade algorithm.

Sam-Irv committed:
Update ai.py

performed changes requested by Jacob
